File: python_server/KAifaAMSParser.py
import struct


class KAifaAMSParser:
    Packet = {}

    NameList = {
        1: ['ActivePowerPositive'],
        9: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative',
            'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'ULN1'],
        13: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative',
             'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'IL2', 'IL3', 'ULN1', 'ULN2', 'ULN3'],
        14: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative', 'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'ULN1',
             'MeterDateTime', 'CumulativeHourlyActiveImport', 'CumulativeHourlyActiveExport', 'CumulativeHourlyReactiveImport', 'CumulativeHourlyReactiveExport'],
        18: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative', 'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'IL2', 'IL3', 'ULN1', 'ULN2',
             'ULN3', 'MeterDateTime', 'CumulativeHourlyActiveImport', 'CumulativeHourlyActiveExport', 'CumulativeHourlyReactiveImport', 'CumulativeHourlyReactiveExport']
    }

    @staticmethod
    def UnpackShort(data: bytes):
        return struct.unpack(">H", data[0:2])[0]

    @staticmethod
    def UnpackInt(data: bytes):
        return struct.unpack(">I", data[0:4])[0]

    @staticmethod
    def ParseDate(data: bytes):
        year = KAifaAMSParser.UnpackShort(data[1:3])
        month = data[3]
        day = data[4]
        hour = data[6]
        minute = data[7]
        second = data[8]
        return {
            'DateTimeString': "{}.{}.{} {}:{}:{}".format(year, str(month).rjust(2, '0'), str(day).rjust(2, '0'), str(
                hour).rjust(2, '0'), str(minute).rjust(2, '0'), str(second).rjust(2, '0')),
            'Weekday': data[5],
            'Hundredths of Seconds': (0 if data[9] == 0xFF else data[9]),
            'Deviation': KAifaAMSParser.UnpackShort(data[10:12]),
            'Clock Status': data[13]
        }

    @staticmethod
    def ParseString(data: bytes):
        if data[0] == 0x09:
            length = data[1]
            return length, str(data[2: length + 2])

    @staticmethod
    def ParseList(data: bytes):
        elementList = list()
        if data[0] == 0x02:
            elements = data[1]
            names = KAifaAMSParser.NameList[elements]

            index = 2
            current = 0
            while current < elements:
                if data[index] == 0x09:
                    length, string = KAifaAMSParser.ParseString(data[index:])
                    elementList.append(
                        {'Name': names[current], 'Value': string})
                    index += length + 2
                elif data[index] == 0x06:
                    elementList.append(
                        {'Name': names[current], 'Value': KAifaAMSParser.UnpackInt(data[index + 1:])})
                    index += 5
                else:
                    print('Unknown key: ', data[index])
                    break
                current += 1
        return elementList

    def ParseHeader(self, data: bytes, index: int):
        header = {}

        header['FrameFormat'] = data[index] >> 4
        header['FrameLength'] = (
            ((data[index] << 7) & 0xF00) >> 1) + data[index + 1]
        index += 2
        checksumIndex = data.index(0x10) + 1
        header['Other'] = str(data[index:checksumIndex])
        header['Checksum'] = KAifaAMSParser.UnpackShort(
            data[checksumIndex:checksumIndex+2])
        self.Packet['Header'] = header
        return checksumIndex + 2

    def ParseBody(self, data: bytes, index: int):
        body = {
            'Destination LSAP': data[index],
            'Source LSAP': data[index + 1],
            'LLC Quality': data[index + 2],
            'LLC Service Data Unit': data[index + 3],
            'Long Invoke ID and Priority': KAifaAMSParser.UnpackInt(data[index + 4:index + 8]),
            'DateTime': KAifaAMSParser.ParseDate(data[index + 9:])
        }
        index += 22
        if(data[index] == 0x02):
            listElements = KAifaAMSParser.ParseList(data[index:])
        body['List'] = listElements
        self.Packet['Body'] = body

    def ParseFooter(self, data: bytes, index: int):
        footer = {
            'Checksum': self.UnpackShort(data.slice(index, index + 2))
        }
        index += 2
        if data[index] != 0x7E:
            # More data than expected?
            overFlowIndex = data.index(0x7E, index)
            footer['Overflow'] = str(data[index:overFlowIndex])
            index += overFlowIndex - index
        self.Packet['Footer'] = footer
        return index

    def Parse(self, data: str):
        index = 0
        if data[index] == 0x7E and data[index + 1] != 0x7E:
            index += 1
            index = self.ParseHeader(data, index)
            self.ParseBody(data, index)

        return self.Packet

File: nodejs_server/KaifaAMSParser.js
module.exports = (function () {
    const BytePack = require('./bytepack');

    let Packet = {}
    const NameList = {
        1: ['ActivePowerPositive'],
        9: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative',
            'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'ULN1'],
        13: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative',
            'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'IL2', 'IL3', 'ULN1', 'ULN2', 'ULN3'],
        14: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative', 'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'ULN1',
            'MeterDateTime', 'CumulativeHourlyActiveImport', 'CumulativeHourlyActiveExport', 'CumulativeHourlyReactiveImport', 'CumulativeHourlyReactiveExport'],
        18: ['OBILListVersionIdentifier', 'MeterID', 'MeterType', 'ActivePowerPositive', 'ActivePowerNegative', 'ReactivePowerPositive', 'ReactivePowerNegative', 'IL1', 'IL2', 'IL3', 'ULN1', 'ULN2',
            'ULN3', 'MeterDateTime', 'CumulativeHourlyActiveImport', 'CumulativeHourlyActiveExport', 'CumulativeHourlyReactiveImport', 'CumulativeHourlyReactiveExport']
    }

    this.ParseDate = function (bytes = []) {
        let year = BytePack.UnpackShort(bytes.slice(1, 3));
        let month = String(bytes[3]).padStart(2, '0');
        let day = String(bytes[4]).padStart(2, '0');

        let hour = String(bytes[6]).padStart(2, '0');
        let minute = String(bytes[7]).padStart(2, '0');
        let second = String(bytes[8]).padStart(2, '0');
        return {
            DateTimeString: `${year}.${month}.${day} ${hour}:${minute}:${second}`,
            Weekday: bytes[5],
            'Hundreths of Seconds': bytes[9] == 0xFF ? 0 : bytes[9],
            Deviation: BytePack.UnpackShort(bytes.slice(10, 12)),
            'Clock Status': bytes[13]
        };
    }

    this.ParseString = function (bytes = []) {
        if (bytes[0] == 0x09) {
            let length = bytes[1];
            return [length, bytes.slice(2, length + 2).map((value) => {
                return String.fromCharCode(value);
            }).join("")];
        }
    }

    this.ParseList = function (bytes = []) {
        let elementList = [];
        if (bytes[0] == 0x02) {
            let elements = bytes[1];
            let names = NameList[elements];

            let index = 2;
            let current = 0;
            while (current < elements) {
                let identifier = bytes[index];
                if (identifier == 0x09) {
                    const [length, string] = this.ParseString(bytes.slice(index));
                    elementList.push({ Name: names[current], Value: string });
                    index += length + 2
                } else if (identifier == 0x06) {
                    elementList.push({ Name: names[current], Value: BytePack.UnpackInt(bytes.slice(index + 1, index + 5)) });
                    index += 5;
                } else {
                    console.log(`Unknown key: ${identifier}`)
                }
                current++;
            }
            return [index, elementList];
        }
        return [0, elementList];
    }

    this.ParseHeader = function (bytes = [], index) {
        Packet.Header = {
            FrameFormat: bytes[index] >> 4,
            FrameLength: (((bytes[index] << 7) & 0xF00) >> 1) + bytes[index + 1],
        }
        index += 2;
        let checksumIndex = bytes.indexOf(0x10) + 1;
        Packet.Header.Other = String(bytes.slice(index, checksumIndex));
        Packet.Header.Checksum = BytePack.UnpackShort(bytes.slice(checksumIndex, checksumIndex + 2));
        return checksumIndex + 2;
    }

    this.ParseBody = function (bytes = [], index) {
        Packet.Body = {
            'Destination LSAP': bytes[index],
            'Source LSAP': bytes[index + 1],
            'LLC Quality': bytes[index + 2],
            'LLC Service Data Unit': bytes[index + 3],
            'Long Invoke ID and Priority': BytePack.UnpackInt(bytes.slice(index + 4, index + 8)),
            DateTime: this.ParseDate(bytes.slice(index + 9))
        }
        index += 22;
        if (bytes[index] == 0x02) {
            const [length, elementList] = this.ParseList(bytes.slice(index));
            Packet.Body.List = elementList;
            index += length;
        } else {
            Packet.Body.List = []
        }
        return index;
    }

    this.ParseFooter = function (bytes = [], index) {
        Packet.Footer = {
            Checksum: BytePack.UnpackShort(bytes.slice(index, index + 2))
        };
        index += 2;
        if (bytes[index] != 0x7E) {
            // More data than expected?
            Packet.Footer.Overflow = String(bytes.slice(index, bytes.indexOf(0x7E, index)));
        }
        return index;
    }

    this.Parse = function (bytes = []) {
        let index = 0;
        if (bytes[index] == 0x7E && bytes[index + 1] != 0x7E) {
            index++;
            index = this.ParseHeader(bytes, index);
            index = this.ParseBody(bytes, index);
            index = this.ParseFooter(bytes, index);
        }
        return Packet;
    }
    return this;
})();